py/server/deephaven/updateby_clause.py

```diff
@@ -0,0 +1,269 @@
+#
+#     Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+"""This module supports building various update-by clauses for use with the update-by operations on tables. """
+from enum import Enum
+from typing import Union, List
+
+import jpy
+
+from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.jcompat import to_sequence
+
+_JUpdateByClause = jpy.get_type("io.deephaven.api.updateby.UpdateByClause")
+_JBadDataBehavior = jpy.get_type("io.deephaven.api.updateby.BadDataBehavior")
+_JEmaControl = jpy.get_type("io.deephaven.api.updateby.EmaControl")
+_JMathContext = jpy.get_type("java.math.MathContext")
+
+
+class MathContext(Enum):
+    UNLIMITED = _JMathContext.UNLIMITED
+    """unlimited precision arithmetic"""
+
+    DECIMAL32 = _JMathContext.DECIMAL32
+    """a precision setting matching the IEEE 754R Decimal32 format, 7 digits"""
+
+    DECIMAL64 = _JMathContext.DECIMAL64
+    """a precision setting matching the IEEE 754R Decimal64 format, 16 digits"""
+
+    DECIMAL128 = _JMathContext.DECIMAL128
+    """a precision setting matching the IEEE 754R Decimal128 format, 34 digits"""
+
+
+class BadDataBehavior(Enum):
+    """An Enum defining ways to handle invalid data during EMA processing."""
+
+    RESET = _JBadDataBehavior.RESET
+    """Reset the state for the bucket to {@code null} when invalid data is encountered"""
+
+    SKIP = _JBadDataBehavior.SKIP
+    """Skip and do not process the invalid data without changing state"""
+
+    THROW = _JBadDataBehavior.THROW
+    """Throw an exception and abort processing when bad data is encountered"""
+
+    POISON = _JBadDataBehavior.POISON
+    """Allow the bad data to poison the result. This is only valid for use with NaN"""
```

Comment:
These look like they are potentially generally useful.  Do they belong specifically in this module?

Reply:
For now these are only used by update-by

Reply:
I understand.  I'm just trying to see if we expect it to stay that way, or if we think there are reasonable odds that they might expand in the future.